### ISSUE TITLE
A new permission for canvas data

### ISSUE THREAD

**Issue by tomrittervg** [w3c/permissions#165]
Canvas can (and is) used to fingerprint users, allowing (for example) ad agencies to track users even when they wipe their cookies and other persistent identifiers and change networks. (More information about this is available at https://en.wikipedia.org/wiki/Canvas_fingerprinting )

The fingerprinting ultimately arises by asking the canvas to render something, and then recording the resulting render using (for example, toDataURL or toBlob). Subtle differences in how the image is rendered can be unique to individual users, and is repeatable across time, network, and domain.

For a long time, Tor Browser has used a Permission Prompt to control access to these APIs; and more recently we (Mozilla) have adopted that patch and integrated it (behind a non-default preference). A downside of being more privacy conscious is that canvas use is prevalent throughout the web, and the permission prompt occurs quite a lot, and it's not possible to distinguish between benign uses of canvas (such as using it to check for emoji support) and tracking purposes.

However, we've spoken with at the maintainer of the Wordpress emoji library (which is used quite a bit around the web) and they are willing to account for the Permission prompt if we can provide a way to query "Will you display a permission prompt" in JavaScript. We hope other large sites would consider doing the same.

Obviously the Permissions API is the best way to query "Will you show a permission prompt?"

So I'd like to standardize a new permission (perhaps 'canvas-pixeldata') that will state whether a UA will prompt for permission to access the canvas pixel APIs (toDataURL, toBlob, and maybe isPointInPath, the last one needs investigating.)

**Comment by jyasskin** [comment]
I'd naively hope that a better approach would be to standardize algorithms for drawing on canvases, so that privacy-sensitive browsers can get results that are uniform across users. Even if the result is slower, that seems like a better result for people using Tor than simply being locked out of canvas-using websites. However, you've almost certainly already tried that, and it doesn't work for some reason. I'd appreciate if you write a Note describing that reason, and include it with the new permission.

I want to ask if you have evidence that users understand the prompt, but actually if Mozilla's shipping a prompt for something, it probably makes sense to add the permission name eagerly so that sites can check for it. @raymeskhoury, what do you think?

@tomrittervg, I assume you'll write the PR for this spec? Are you also volunteering to write the change to the [canvas spec](https://html.spec.whatwg.org/multipage/canvas.html) to use the "canvas-pixeldata" permission?

**Comment by isislovecruft** [comment]
Hi! I [worked on](https://trac.torproject.org/projects/tor/ticket/12684) the phrasing/UI of the canvas fingerprinting prompt a tiny bit in Tor Browser. Also caveat lector: I'm not a UX person or a web person _at all_, so my apologies for jumping in, and also if I misunderstand anything.

@jyasskin System/platform independent ways would be an excellent eventual goal! It would be amazing if this were something we could all collectively aim for. However, (afaik, but I've not kept up on the last couple years of research) it's not even entirely understood if we _can_ attain system independent rendering. In any case, it would be a long way away (and not to mention continually dependent upon several vendors _continuing_ to work together to maintain mutual conformity).

Part of the problem that @tomrittervg has described, and which would be great to fix on a sooner timeline, boils down to this: when a site attempts to render a canvas, it doesn't check first (it doesn't currently have any way to!) whether it's allowed to do so. Our solution at The Tor Project was, upon detecting a render attempt, to simply return a blank, white canvas and notify the user that the attempt had occurred, along with asking their permission for the future. (@tomrittervg can probably clarify/correct, but I believe Mozilla is using the same strategy.) However, the first attempt will fail. This causes users to get confused and frustrated: for example, someone is trying to set their Twitter profile picture (they spend some time getting the framing/size just right, etc.), but they get an error message from Twitter due to Twitter receiving a blank canvas in return. (After they did all that work to get it just right too!) They then need to 1) allow the permission and then 2) try setting their picture again.  If we had the "ask for permission" mechanism that @tomrittervg is describing, the flow would hopefully be something more like 1) the site wishes to render a canvas 2) it asks the user's permission 3) the user gives permission 4) the canvas is rendered and sent.

**Comment by raymeskhoury** [comment]
Having a prompt to display a Canvas is very bad from a UX perspective. But I understand the need if you're optimizing for privacy.

A part of me wonders if we should have a more general fingerprinting permission that can be used for other things too..

**Comment by jyasskin** [comment]
@raymeskhoury I thought about suggesting the more general fingerprinting permission, but I think the developer-facing API winds up being harder to use as we cover new APIs with it. For example, say we've rolled out the Canvas.toBlob() prompt under "fingerprinting", and then we want to add a prompt for (making something up) token-bound fetching. How should a site figure out that the browser's going to prompt for their fetch?

I think the prompt itself (e.g. "can this site do things that might fingerprint you?") could grant several of these permissions at once, like we were considering for the movement sensors.

@isislovecruft I'm understanding a different request from your post than from @tomrittervg's post. Specifically, I believe Tom asked for a way for the site to ask "if I do this, will a prompt appear?", while I think you're asking for a way for a site to ask "please show the prompt now, if you would at all." Tom's is what the [permissions.query()](https://w3c.github.io/permissions/#dom-permissions-query) function in this spec does, while yours is handled by [permissions.request()](https://jyasskin.github.io/permissions-request/), which got kicked out to another spec at the request of some other Mozilla folks.

They're both sensible things to want ... and while [toBlob()](https://html.spec.whatwg.org/multipage/canvas.html#dom-canvas-toblob) is fine with just the query, now that I've looked at [toDataURL()](https://html.spec.whatwg.org/multipage/canvas.html#dom-canvas-todataurl) more, because it returns a result synchronously, it'll probably still have to return a blank result while query() returns "prompt". This is *probably* ok, since site developers who are aware of the problem can call toBlob() instead in order to wait for the user to answer the prompt. request() would be a more explicit way to do this, but it's not essential.

In summary, I'm currently +1 on adding the "canvas-pixeldata" PermissionName for use with permissions.query().

**Comment by KOLANICH** [comment]
I'm against letting the apps  know if we show a prompt if they will fingerprint (unless we always show the prompt). Because letting them know may alter their behavior. So they will fingerprint only the ones who won't notice that. This will make fingerprinting more stealthy , the web less transparent and people more ignorant.

**Comment by annevk** [comment]
Bikeshed: if we do this (@KOLANICH brings up an interesting point I had not considered) I suggest "canvasimagedata" or "canvas-imagedata" if hyphens are preferred. That would match existing API terminology more closely.

(As for prompting for synchronous APIs, I still hold out some hope that eventually we know enough about what things we are executing so we don't have to return to script until we've heard from the user.)

**Comment by tomrittervg** [comment]
First off, yes, getting canvas to be unfingerprintable is definitely desirable. But we've known about this problem for years and no browser has actually done it. And as long as browsers drive towards GPU-based graphics acceleration, it seems likely that it won't happen. Convincing people to go backwards to software rendering (which addresses some but not all all issues) is... not an easy thing to convince people of, at least for a default mode.

I'm definitely willing to do the spec work, yes. My suspicion is that documenting the Privacy Considerations (and why it's difficult to make canvas unfingerprintable) will be an extended effort. Mostly because I'd like it to be as complete as possible, and it's likely there are even more concerns that have come up since this was last investigated years ago that will take some time to discover.

---

As for the canvas spec. As Isis noted, because Canvas uses synchronous APIs the current behavior in both Tor Browser and Firefox is... cruddy. Prompts get thrown all over the web (see the emoji note) even when fingerprinting isn't occurring; and not all websites can handle the non-standard, unexpected situation of "The browser just lied to you with a blank canvas, but if you call the API again it will work this time!" For the ones that do, it's more luck than robust design.

The ideal situation would be for the APIs of concern to be Promise-based. I am not asking for that in this request. I am basically asking for a string allocation to standardize the current cruddy behavior. I'd be willing to update the Canvas spec to document this possible behavior, but it's definetly a hack. @annevk's notion about pausing execution is very promising, but I'd strongly prefer not to hinge development of solutions here on it.

Now in the future, if canvas changed, this string we allocate would easily accommodate a new, not-cruddy behavior! And if people think it would be well received over in WHATWG I could propose that - but I worry it wouldn't get much traction. It'd basically be converting toDataURL to be promise-based (since toBlob is callback-based).  (There is a snag: isPointInPath looks like it would leak data as well, but that is an ugly API to turn asynchronous. We'd have to investigate it more to learn how much data you can extract from it.) And of course, the web would just need to migrate to the new APIs for better behavior ;)

Separately, there is a new(ish?) [OffscreenCanvas thing](https://html.spec.whatwg.org/multipage/canvas.html#offscreencanvas). It has two APIs: convertToBlob (which is Promise-based already!) and transferToImageBitmap (which is synchronous). Making the latter asynchronous would be awesome, and maybe it would get more traction because it's a newer thing? (Or at least, the web hasn't deployed it everywhere yet?)

---

Separately, I don't like the idea of a general "Can this site fingerprint you?" permission prompt. My hope is that some of the many fingerprinting vectors will be reconsidered in a variety of browsers. For example: so few websites actually use a gamepad, and the tail of users who have gamepads (and which model) is so small; GamePads seems like something that could reasonably be prompted for, as opposed to letting every website on the Internet get that information right away. Similarly, I have a very theoretical idea about making a [prompt for Timezone](https://bugzilla.mozilla.org/show_bug.cgi?id=1426232).

I'm not trying to talk about either of those in this thread, I'm just using them to demonstrate that in the future I hope multiple browsers will consider prompts for fingerprinting resistance of different APIs, and having a single permission that implies A, B, and C in one browser, B and C in another browser, and C in an earlier version of that browser is a bad situation to be in spec and compatibility-wise.

---

> I'm against letting the apps know if we show a prompt if they will fingerprint (unless we always show the prompt). Because letting them know may alter their behavior. So they will fingerprint only the ones who won't notice that. This will make fingerprinting more stealthy , the web less transparent and people more ignorant.

If a website *wants* to be stealthy or malicious, it's not difficult to know how the browser will behave. It's just side-channel feature detection ('Oh you're Firefox 32, then you must support foo'), something the web did for years and now we're working to deprecate. Supporting a simple permission query will make it easy for the web to learn this information in a way that is future-proof. And it will make the web experience for a resist fingerprinting mode better, and more likely to get such an experience into browsers by default.

Here's a simple algorithm to know if a browser will show a permission prompt on canvas. It's not 100% accurate, but it's going to have a very low false positive and a very low false negative rate.

1. Is the client IP address a tor exit node, using [TorDNSEL](https://www.torproject.org/projects/tordnsel.html.en) -> If Yes: It will show a prompt
2. Is the browser User Agent Firefox, the version 52 (or 60), and is the result of performance.now() rounded exactly to 100ms -> If Yes: It will show a prompt

**Comment by jyasskin** [comment]
@annevk 

@tomrittervg It seems like it'd be fine for the TOR browser if its host offered a software (or otherwise consistent) rendering path, which TOR could opt into, even if that host browser didn't make that the default for non-TOR users. This wouldn't help you blend into the larger non-TOR population, but as you said, sites can already retrieve those bits of identifying entropy.

That said, I'm not insisting that you have that argument with the Mozilla rendering folks before adding the PermissionName.

----

I don't expect @annevk to be successful in pausing execution while a user responds to a permission prompt. 

I do think there's appetite to add an asynchronous variant of Canvas.toDataURL() (and [getImageData()](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-getimagedata) and any other synchronous image-reading functions): it handles bulk data that we might prefer to live on another thread, on the GPU, or even on disk. Tab Atkins and Ojan Vafai are excited about this idea.

There's probably similar interest in making isPointInPath() asynchronous, since evaluating it could be expensive.

I *think* that  OffscreenCanvas.transferToImageBitmap() is safe to leave synchronous because the ImageBitmap doesn't actually provide any pixel data: it's just a handle to put the pixels somewhere else.

----

I think @KOLANICH's worry that query("canvas-imagedata") will be a net negative for fingerprinting is the same as the worry in #52 that allowing sites to know whether a prompt will happen, will allow them to more sneakily gather private data. We've basically rejected that concern, although we weren't explicit about it in that issue.

**Comment by KOLANICH** [comment]
>even when fingerprinting isn't occurring

But it is really occuring. Detection of emojis present in the system is in fact font fingerprinting. Often the goal of that fingerprinting is not to identify a user, but this problem is unsolvable without a general-purpose AI. So, from a point of view of a dumb computer it is fingerprinting and we cannot do anythong about it for now.

**Comment by KOLANICH** [comment]
>Supporting a simple permission query will make it easy for the web to learn this information in a way that is future-proof.

Permission query is nice, but the app in an ideal world  should be unable to know if the permission is really granted, or if it was revoked, but faked. See also #54.

>is the result of performance.now() rounded exactly to 100ms

it's a separate fingerprint vector, which definitely must be closed by rounding and then adding random noise uniformly distributed in the interval [-50, +50].

**Comment by pento** [comment]
I'm the WordPress maintainer mentioned in the OP. 🙂

We've found over time that the canvas render method is the only reliable way to determine if the current browser/OS will render emoji correctly. There's some relevant discussion on [one of our tickets](https://core.trac.wordpress.org/ticket/42428).

I certainly appreciate that the privacy issues are more complex than simply providing a flag to determine if the current browser will allow the canvas image data or not. If something resembling that flag were available, however, we'd definitely want to include a check for it on all WordPress sites.

An alternative option for us, which is wildly out of scope for this issue (but may be of interest), is if the browser was able to determine if it supports rendering current emoji sets, we'd be more than happy to use that information instead. We've found that we can split emoji support into testing for whether we can render all flags, and whether we can render other emoji from the latest spec. Currently, we test the [UN](https://emojipedia.org/flag-for-united-nations/) and [England](https://emojipedia.org/flag-for-england/) flags, and the [Male Fairy](https://emojipedia.org/man-fairy/) emoji.

**Comment by annevk** [comment]
Inspecting what code points the browser can find a glyph for basically has the same privacy implications. Though given that you can determine it through rendering perhaps we should offer some kind of more straightforward API around that (and make it async so it can be blocked more easily)...

@tabatkins has the CSS WG been thinking about such a thing?

**Comment by KOLANICH** [comment]
I vote agains't an API to inspect glyphs and get their parameters. I insist on that the existing fingerprinting vectors must be eliminated, including [measureText and getBoundingClientRectangle-based](https://github.com/KOLANICH/Article-2015-Dull-captaincy-or-the-way-Tor-Project-fights-browser-fingerprinting). For example it is possible try to make the text rendering identical on all the systems, shipping the fonts, maybe even a separate pack of fonts for each operation system targeted to compensate the differrences between rendering. Then you can decouple results of text-measuring API from the real results and just hope that they correspond. If they don't, c'est la vie, someone's rendering got a little broken.

**Comment by annevk** [comment]
If you ship identical fonts an API that inspects whether a code point is backed by a glyph would give platform-independent results, no?

**Comment by KOLANICH** [comment]
Not quite. But I guess it possible to check the glyphs. But it's a one more fingerprint vector: unicode constantly updates, people want to get the latest glyphs, so the version of browser will be fingerprintable using glyphs. So I'm against the api to check glyphs.

The idea is not only to ship fonts, but to take a base font and measure its glyphs on every supported environment. So we get a tensor n environments * m glyphs in superset of glyphs all the used fonts * 2 (width and height). m_{i,j,k}, i is env, j is glyph , k is dimension indexes.

Then we compute the median over each glyph in measurements and get Me_{j, k}.
This table is serialized, standardized and shipped with each browser. When a user starts a browser, a browser does rendering and calculates the sizes of glyphs the same way the reference ones are obtained. Then it calculates the differences. Then it modifies the shipped font files in the direction which should eliminate the difference (gradient descent). Then repeats until convergence. Then it does the same with ligatures and kerning. After that the rendered font glyph sizes are considered similar enough to standardized ones. So for measuring we can return not the actual sizes, but the predicted ones. The prediction algorithm must be standardized, so every browser must return the same value. And this should not make much disturbance, because simulated dimensions should be close enough to the real ones. This is only relevant to glyph size fingerprinting I have used in that research. Canvas image data still mustn't be exposed.

**Comment by jyasskin** [comment]
1. A new glyph-detecting API seems out of scope for this repository and specification.
1. It's easy to use https://wpt.fyi/ to find test cases that distinguish just about any two browser versions, so I stop listening when the justification for a change is "prevent people from distinguishing browser versions". It only makes sense to hide in a crowd consisting of people using the same version of the same browser as you.
1. @tomrittervg's and @isislovecruft's goals for Tor still seem reasonably well served by a new `PermissionName` here, unless they want to rescind that request for now in favor of trying to get a GPU-independent path added to Firefox. This can proceed independently of an asynchronous replacement of the synchronous canvas-reading APIs. (The asynchronous APIs can improve the UX of requesting permission, but they don't replace sites' need to know whether a prompt will show at all.

**Comment by tomrittervg** [comment]
I'm going to pass on the discussion of default fonts (and per-install font-rendering-munging) except to say we're going to do the former (https://bugzilla.mozilla.org/show_bug.cgi?id=1336208) and the latter sounds interesting but very complicated.

We're going to investigate GPU-independent paths, but I'm still interested in the permission name.

**Comment by tabatkins** [comment]
> Inspecting what code points the browser can find a glyph for basically has the same privacy implications. Though given that you can determine it through rendering perhaps we should offer some kind of more straightforward API around that (and make it async so it can be blocked more easily)...
>
> @tabatkins has the CSS WG been thinking about such a thing?

More or less - Houdini is planning on including a much more robust font-metrics API, so you can do complicated font-specific rendering on your own (like what the 'initial-letter' property does, for example).

Privacy implications are minimal, since the horse has already left the barn - as you say, it's quite trivial to probe the list of a user's installed fonts today by just inspecting layout (and this apparently gives quite a few bits of entropy when fingerprinting).

**Comment by KOLANICH** [comment]
>Privacy implications are minimal, since the horse has already left the barn - as you say, it's quite trivial to probe the list of a user's installed fonts today by just inspecting layout (and this apparently gives quite a few bits of entropy when fingerprinting).

This must be eliminated disregarding compatibility. If any website uses this, its should be their problem. Our problem is to eliminate the fingerprinting vector without crippling non-fingerprinting uses. I'm not CSS spec, so I wonder if getBoundingClientRect can be replaced with CSS entirely. I mean if it is possible implement and enforce the rule "if a one needs getBoundingClientRect to compute complex layout, he must use purely CSS solution rather than use JS for it, and the sizes mustn't leak to JS".  Of course, in fact this initiative may be blocked by some companies taking major share of browser market and supporting fingerprinting.

**Comment by tabatkins** [comment]
> This must be eliminated disregarding compatibility. If any website uses this, its should be their problem. Our problem is to eliminate the fingerprinting vector without crippling non-fingerprinting uses.

The browser can handle that, as mentioned in the Firefox bug up above, by whitelisting a set of local fonts visible to the webpage. That avoids the fingerprinting vector entirely, and I'd support that. (The page would then only have access to the whitelist fonts, and whatever they pull in themselves via @font-face.)

> I'm not CSS spec, so I wonder if getBoundingClientRect can be replaced with CSS entirely. I mean if it is possible implement and enforce the rule "if a one needs getBoundingClientRect to compute complex layout, he must use purely CSS solution rather than use JS for it, and the sizes mustn't leak to JS". 

It can't be.  CSS, while powerful, has nothing on the abilities of a full programming language like JS, and gBCR can't generally be replaced.  This would basically kill *all* alternative layouts besides the few that CSS explicitly gives you.

> Of course, in fact this initiative may be blocked by some companies taking major share of browser market and supporting fingerprinting.

This is neither true nor kind, so it fails the conversational test.

**Comment by KOLANICH** [comment]
>The browser can handle that, as mentioned in the Firefox bug up above, by whitelisting a set of local fonts visible to the webpage. That avoids the fingerprinting vector entirely, and I'd support that. (The page would then only have access to the whitelist fonts, and whatever they pull in themselves via @font-face.)


According to my previous (long ago) experiments even identical fonts and browser binaries leak some info about environment. I had posted the link above, you can try that yourself on your PC. First install TBB into a main system and acquire a fingerprint. Then fully restart TBB and acquire another fingerprint and make sure that they match. Then run an OS in a VM, install TBB there and aquire another fingerprint. Then compare the fingerprints. You may also want to try it with different versions of usual Firefox with privacy.resistFingerprinting=1.

**Comment by jackpalmer1** [comment]
Is this going to be a thing be for TOR or mainstream browsers? I dont think a permission is required for this thing because no personal information is leaked. Maybe this only fits TOR, and we already block TOR nodes due to spam and abuse. Even ClouldFlare and Google's ReCapcha dislikes TOR and block them most of the time. 

Having excessive permission for even rendering and getting rendered data to be displayed in some other form is going to cause a permission fatigue. There is no critical personal permission involved in this case.  

TOR probably has different priorities than regular browser and they come at a huge cost. Sadly we've invested too much time in developing a PWA to move away from native our native app on phones, without realizing that we are at the mercy of browser vendors. 

We already stopped testing on Firefox because of this, because we're not going to deal with edge cases that get introduced after 3 years into development of our product when all of a sudden Mozilla decides it, and much of our spam comes from Firefox and TOR users. After they introduced it we started shadow banning Firefox users who present us with a white canvas, because spammers hide in legitimate users. If Firefox starts doing this by default, I'm afraid they'll get a sub par treatment like TOR users. We're not going to fingerprint Firefox users, but we're not going to allow allow them on the platform if this happens because Facebook already has a very negative reputation after Cambridge Analytica fiasco and there is a stigma about privacy even though there is not much of a risk here. 

I'd like to know Google's stance on this before we decide to launch our desktop product pwa. We already had a redesign, but there are further improvements that are being made for better visual performance. If this is also on their roadmap than we'll publish on the Windows Store instead of the web and be done with it. We use fingerprinting for bot detection and avoid multiple signups for vote manipulation. or shadow-banning spammers. And to be honest the fingerprint is only accurate until the next minor release of a browser. Of course there is not enough information to create a unique id but it is good enough to hold information for next 48 hours and disallow signups from same users while not informing them. If this happens, then we can expect that permissions could be introduced for things that don't require permission now, and we would move to a native app on Windows entirely. The web unfortunately is already prone to abuse, and since our website is in Alexa Top 10, we can no longer allow this. In-fact there are already many attempts for vote manipulation, and its is difficult to keep a watch on suspicious users. 

And this only works because most legitimate users dont know about this. Spammers and very few privacy users fake their fingerprint, legitimate users do not. and as a result spammers and vote manipulators get shadow banned. 

It would be helpful if browsers could provide us with a standard device id to avoid fraud and manipulation, because the other way is to force them to download a native app which we don't want to but have to for reasons of the safety of the community as a whole. Since fingerprint on mobile are not unique enough, we have to partner with several apps that are already likely to be on the users device to perform the risk analysis in order to decide shadow banning. 

We would like to once again remind that if it is decided for mainstream browser, drawbacks and benefits need to be taken into account. Many banking apps use the same thing for detecting fraud, and Akamai and other folks likely do the same thing for spam detection. 

Also fingerprint it not deterministic but probabilistic that it allows for risk analysis.

**Comment by KOLANICH** [comment]
>Maybe this only fits TOR, and we already avoid TOR exit nodes due to spam and abuse. Even ClouldFlare and Google's ReCapcha dislikes TOR and block them most of the time.

Keep avoiding. You have chosen the wrong place to promote your views. You could go to either congress or parliament, I'm sure some of these folks will be interested in promoting mass surveillance.

>Having excessive permission for even rendering and getting rendered data to be displayed in some other form is going to cause a permission fatigue

This should be personal troubles of the ones wanna track. I don't think that W3C should help them.

>We already stopped testing on Firefox because of this

Again, you can say  "just use chrome or f*** o*" to your potentia, users. It's completely OK. You can also ask them to install any kind of DRM and malware you want. I don't think that Mozilla should help trackers to track its users.

**Comment by jackpalmer1** [comment]
reddit

**Comment by photopea** [comment]
I understand, that you can do fillText() / measureText() on Canvas, get pixel data and detect, which fonts are available in the device, and use these fonts as a fingerprint.

But I don't understand, why are you talking specifically about Canvas (you even have it in the title of this issue). The "attacker" can make a <span>Hello</span> element, set various fonts with CSS, and measure the result using Element.getClientBoundingRect(), and get the list of fonts this way.

I wonder, why so much effort is put into Canvas "protection", while nobody cares about getClientBoundingRect() . Using this logic, the website should ask for premissions, when a webpage wants to use the Arial font for its text.

I propose the following solution: Render the text (HTML, Canvas, SVG, ...) **only** when the webpage provides the necessary fonts. I.e. don't render any text, when a webpage provides no fonts. 

In addition, the w3c can define a list of standard fonts, that have to be present in every web browser (to be used by websites, which don't provide their own fonts). E.g. we can use ["14 PDF fonts"](http://www.enfocus.com/manuals/ReferenceGuide/PP/10/enUS/en-us/concept/c_aa1140975.html) (fonts, that must be present in every PDF renderer). 

Restricting browsers in this way would guarantee the same look of a website on every device, and would remove all "fingerprint vulnerabilities".

**Comment by KOLANICH** [comment]
>But I don't understand, why are you talking specifically about Canvas (you even have it in the title of this issue). The "attacker" can make a Hello element, set various fonts with CSS, and measure the result using Element.getClientBoundingRect(), and get the list of fonts this way.
>nobody cares about getClientBoundingRect() .


In fact the PoC by the link above uses that too and probably the protection may help to deal with it. Higher in this thread you may see some ideas for countermeasures, like font fine-tuning for environment. The idea was also reported to bugzilla.mozilla.org, don't know if they gonna implement it, but some staff have subscribed to it.

>Render the text (HTML, Canvas, SVG, ...) only when the webpage provides the necessary fonts. I.e. don't render any text, when a webpage provides no fonts. E.g. we can use "14 PDF fonts" (fonts, that must be present in every PDF renderer).

This will likely not work (the last time I tested, it haven't: differrent PCs (both Windows 7, 32 bit) with the same version of Tor Browser had differrent fingerprints) for the case of Canvas. In fact Tor Browser (and Firefox, if I remember right) bundles fonts. 

About the standard on browser indistinguishability: there is still no such, even though it was proposed long ago.

**Comment by photopea** [comment]
> This will likely not work (the last time I tested, it haven't: differrent PCs (both Windows 7, 32 bit) with the same version of Tor Browser had differrent fingerprints) for the case of Canvas. In fact Tor Browser (and Firefox, if I remember right) bundles fonts.

What different fingerprints of the Canvas do you mean? I think different available fonts are the only fingerprints.

What is font bundling and how is it related to this?

**Comment by KOLANICH** [comment]
@photopea, please read https://cseweb.ucsd.edu/~hovav/dist/canvas.pdf

**Comment by photopea** [comment]
I read your article. They fingerprint 2d context only by rendering fonts. I see, that Linux does not have "Arial" font, so it uses a font, that is a lot different. I also think, that Mac OS uses Helvetica for rendering Arial.

The rendered font differs only for a each combination of  OS and a version of a browser. But this information can be achieved otherwise, without Canvas. So allowing only page-supplied fonts to be rendered should solve all this.

**Comment by KOLANICH** [comment]
@photopea have you read carefully both papers? It's not only about fonts, it's also about software taking part in font rendering. Different versions of libfreetype, DirectDraw, OS : these all affects this.

**Comment by marcoscaceres** [comment]
Closing this issue due to inactivity... I think this might fall into the "too hard" basket for now, but happy to reopen if new information/more-motivation becomes available.
